From 3a294728fd1c7941d83aa76e071a66eaac704f9c Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 15 Sep 2021 07:36:32 +0300
Subject: [PATCH] [SW-54874] gaudi2: arc: load preboot to sram base

Currently, for secureboot flow preboot is loaded at 0x4000 keeping
is similar to btl flow. However in secureboot flow, ppboot is loaded
in FW SRAM instead of SRAM which keeps first 16KB SRAM unused. This
patch add changes to load and execute preboot to SRAM base.

Change-Id: Ib036a411697f7fc948f99d1f32ca6875a048a479
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 6ed11dd..087a8ca 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -205,21 +205,17 @@
 
 	/*
 	 * Copy preboot image from flash to SRAM.
-	 * Note that the offset to be provided to perform DMA must be
-	 * in granuality of 4B. ie. In order to write at
-	 * 0x1000FFFFFD004000ull, the address to be passed is
-	 * 0x1000FFFFFD001000ull instead of 0x1000FFFFFD004000ull
-	 * (right shift offset 0x4000 by 2).
+	 * Note that the offset above base address to perform DMA
+	 * must be passed in granuality of 4 bytes.
 	 */
 	dma_copy_image(image_size, flash_offset,
-			(SDRAM_FULL_BASE + (PPBOOT_SIZE >> 2)),
-			SPI_DMA_DST_HBW);
+			SDRAM_FULL_BASE, SPI_DMA_DST_HBW);
 	if (dma_wait_finish() == DMA_CRC_ERROR)
 		return ARCBOOT_BRINGUP_ERROR;
 
 	image_size -= SECUREBOOT_IMG_HEAD_FOOTER;
 	if (secureboot_verify_image(image_size,
-			(SDRAM_BASE + PPBOOT_SIZE + IMG_LEN_SIZE))
+			(SDRAM_BASE + IMG_LEN_SIZE))
 			== SECUREBOOT_VERIFY_FAILED)
 		return ARCBOOT_BRINGUP_ERROR;
 
diff --git a/arch/gaudi2/arc/common.h b/arch/gaudi2/arc/common.h
index 7df71d3..8aa4ce4 100644
--- a/arch/gaudi2/arc/common.h
+++ b/arch/gaudi2/arc/common.h
@@ -42,7 +42,7 @@
 #define PREBOOT_START_OFFSET	0x8
 #define PPBOOT_START_OFFSET	0xC
 #define PPBOOT_TEXT_BASE	(FW_SRAM_BASE + PPBOOT_START_OFFSET)
-#define PREBOOT_TEXT_BASE	(SDRAM_BASE + PPBOOT_SIZE + PREBOOT_START_OFFSET)
+#define PREBOOT_TEXT_BASE	(SDRAM_BASE + PREBOOT_START_OFFSET)
 
 #define SP_SRAM_BOOT_INFO_OFFSET	0x1800
 #define BOOT_INFO_BASE		(SP_SRAM_BASE + SP_SRAM_BOOT_INFO_OFFSET)
